shell/platform/android/io/flutter/embedding/android/FlutterView.java

```diff
@@ -421,6 +448,75 @@ protected void onSizeChanged(int width, int height, int oldWidth, int oldHeight)
     sendViewportMetricsToFlutter();
   }
 
+  @Override
+  protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
+    super.onLayout(changed, left, top, right, bottom);
+    setWindowManagerDisplayFeatures();
+  }
+
+  @Override
+  protected void onAttachedToWindow() {
+    super.onAttachedToWindow();
+    windowManager.registerLayoutChangeCallback(ContextCompat.getMainExecutor(getContext()),
+        windowManagerListener);
+    windowManager.registerDeviceStateChangeCallback(ContextCompat.getMainExecutor(getContext()),
+        windowManagerDeviceStateListener);
+  }
+
+  @Override
+  protected void onDetachedFromWindow() {
+    windowManager.unregisterLayoutChangeCallback(windowManagerListener);
+    windowManager.unregisterDeviceStateChangeCallback(windowManagerDeviceStateListener);
+    super.onDetachedFromWindow();
+  }
+
+  /**
+   * Refresh {@link androidx.window.WindowManager} and {@link android.view.DisplayCutout}
+   * display features. Fold, hinge and cutout areas are populated here.
+   */
+  private void setWindowManagerDisplayFeatures() {
+    List<DisplayFeature> displayFeatures = windowManager.getWindowLayoutInfo().getDisplayFeatures();
+    List<FlutterRenderer.DisplayFeature> result = new ArrayList<>();
+
+    // Data from androidx.window.WindowManager display features. Fold and hinge areas are
+    // populated here.
+    for (DisplayFeature displayFeature: displayFeatures){
+      Log.v(
+          TAG,
+          "WindowManager Display Feature reported with bounds = "
+              + displayFeature.getBounds().toString()
+              + " and type = "
+              + displayFeature.getType());
+      DisplayFeatureType type = DisplayFeatureType.UNKNOWN;
+      if (displayFeature.getType() == DisplayFeature.TYPE_FOLD) {
+        type = DisplayFeatureType.FOLD;
+      } else if (displayFeature.getType() == DisplayFeature.TYPE_HINGE) {
+        type = DisplayFeatureType.HINGE;
+      }
+      // In next versions of WindowManager, each DisplayFeature has its own posture.
+      // We're making the engine changes as close as possible to that for now.
+      final int posture = windowManager.getDeviceState().getPosture();
+      result.add(new FlutterRenderer.DisplayFeature(displayFeature.getBounds(), type, posture));
+    }
+
+    // Data from the DisplayCutout bounds. Cutouts for cameras and other sensors are
+    // populated here.
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+      WindowInsets insets = getRootWindowInsets();
+      if (insets!=null) {
+        DisplayCutout cutout = insets.getDisplayCutout();
+        if (cutout!=null) {
+          for (Rect bounds: cutout.getBoundingRects()){
```

Comment:
nit, `for (Rect bounds : cutout.getBoundingRects()) {` for formatting consistency